lib/galaxy/tools/parameters/__init__.py

```diff
@@ -312,7 +312,82 @@ def populate_state(request_context, inputs, incoming, state, errors={}, prefix='
     3
     >>> print(state['b'][0]['d'][0]['f']['h'])
     4
+    >>> # now test with legacy_param_form=False
+    >>> flat_new = {'a': 1, 'b': [{'c': 2, 'd': [{'e': 3, 'f': {'h': 4, 'g': True}}]}]}
+    >>> state_new = OrderedDict()
+    >>> populate_state(trans, inputs, flat_new, state_new, check=False, legacy_param_form=False)
+    >>> print(state_new['a'])
+    1
+    >>> print(state_new['b'][0]['c'])
+    2
+    >>> print(state_new['b'][0]['d'][0]['e'])
+    3
+    >>> print(state_new['b'][0]['d'][0]['f']['h'])
+    4
+
     """
+    if legacy_param_form:
+        _populate_state_legacy(request_context, inputs, incoming, state, errors, prefix, context, check, simple_errors)
+        return
+
+    context = ExpressionContext(state, context)
+    for input in inputs.values():
+        state[input.name] = input.get_initial_value(request_context, context)
+        group_state = state[input.name]
+        if input.type == 'repeat':
+            if len(incoming[input.name]) > input.max or len(incoming[input.name]) < input.min:
+                errors[input.name] = 'The number of repeat elements is outside the range specified by the tool.'
+            else:
+                del group_state[:]
+                for rep_index, rep in enumerate(incoming[input.name]):
+                    new_state = {'__index__' : rep_index}
+                    group_state.append(new_state)
+                    populate_state(request_context, input.inputs, rep, new_state, errors, context=context, check=check, simple_errors=simple_errors, legacy_param_form=False)
+
+        elif input.type == 'conditional':
+            # if input.value_ref and not input.value_ref_in_group:
+            #     test_param_key = prefix + input.test_param.name
+            # else:
+            #     test_param_key = group_prefix + input.test_param.name
```

Comment:
I don't understand what this `value_ref` attribute does. Is it safe to remove these lines?